posix: Implement mq_notify()

awojasinski opened the pull request:
The function was the last missing piece of the `_POSIX_MESSAGE_PASSING`
option group. Due to lack of signal subsystem in the Zephyr RTOS
the `sigev_notify` member of the `sigevent` structure that describes
the notification cannot be set to `SIGEV_SIGNAL` - this notification
type is not implemented, the function will return -1 and set `errno`
to `ENOSYS`.

`mq_notify` documentation:
https://pubs.opengroup.org/onlinepubs/9699919799/functions/mq_notify.html

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/66958

awojasinski committed:
tests: posix: Add test cases for mq_notify()

Adds test cases testing following features:
- function error handling
- basic notification type
- thread notification type
- adding notification to non-empty queue

Signed-off-by: Adam Wojasinski <awojasinski@baylibre.com>

awojasinski committed:
posix: Add implementation of mq_notify() function

The function was the last missing piece of the `_POSIX_MESSAGE_PASSING`
option group. Due to lack of signal subsystem in the Zephyr RTOS
the `sigev_notify` member of the `sigevent` structure that describes
the notification cannot be set to `SIGEV_SIGNAL` - this notification
type is not implemented, the function will return -1 and set `errno`
to `ENOSYS`.

`mq_notify` documentation:
https://pubs.opengroup.org/onlinepubs/9699919799/functions/mq_notify.html

Fixes #66958

Signed-off-by: Adam Wojasinski <awojasinski@baylibre.com>

ycsin reviewed (changes requested):
try not to modify unrelated lines, with the right config I think you can format just the modified lines in vscode

awojasinski commented:
> try not to modify unrelated lines, with the right config I think you can format just the modified lines in vscode

@ycsin just tried to fix some formatting issues but don't need to do that

ycsin commented:
Looks like you need a rebase

cfriedt commented on doc/services/portability/posix/option_groups/index.rst:
```diff
@@ -359,7 +359,7 @@ _POSIX_MESSAGE_PASSING
     mq_close(),yes
     mq_getattr(),yes
-    mq_notify(),
+    mq_notify(),yes
```
Is it possible to get the doc update in a separate, 3rd commit?

awojasinski replied:
any preference? Should be first or last?

cfriedt replied:
@awojasinski - last ideally (3rd) - thanks!

awojasinski replied:
@cfriedt done

awojasinski committed:
doc: posix: Update POSIX supported API documentation

Updates in documentation support for `mq_notify` API
in `_POSIX_MESSAGE_PASSING` group option.

Signed-off-by: Adam Wojasinski <awojasinski@baylibre.com>

cfriedt commented on tests/posix/common/src/mqueue.c:
```diff
@@ -97,4 +96,205 @@ ZTEST(mqueue, test_mqueue)
+ZTEST(mqueue, test_mqueue_notify_errors)
+{
+	mqd_t mqd;
+	struct mq_attr attrs;
+	struct sigevent not;
+	int32_t mode = 0777;
+	int flags = O_RDWR | O_CREAT;
+
+	attrs.mq_msgsize = MESSAGE_SIZE;
+	attrs.mq_maxmsg = MESG_COUNT_PERMQ;
+
+	not.sigev_notify_function = notify_function_basic;
+	not.sigev_notify_attributes = NULL;
+	not.sigev_notify = SIGEV_SIGNAL;
+	not.sigev_value.sival_ptr = (void *)&notification_executed;
```
Not a blocker, but `attrs` and `not` could be initialized where they are declared (we support C11 by default, so no need to use only C89 declarations). It might be a bit less verbose, but also, there would be no need to initialize anything to 0 / `NULL` explicitly.
```suggestion
	struct mq_attr attrs = {
		.mq_msgsize = MESSAGE_SIZE,
		.mq_maxmsg = MESG_COUNT_PERMQ,
	};
	struct sigevent not = {
		.sigev_notify_function = notify_function_basic,
		.sigev_notify = SIGEV_SIGNAL,
		.sigev_value.sival_ptr = (void *)&notification_executed,
	};
	int32_t mode = 0777;
	int flags = O_RDWR | O_CREAT;
```

cfriedt commented on tests/posix/common/src/mqueue.c:
```diff
@@ -97,4 +96,205 @@ ZTEST(mqueue, test_mqueue)
+
+ZTEST(mqueue, test_mqueue_notify_thread)
+{
+	mqd_t mqd;
+	struct mq_attr attrs;
+	struct sigevent not;
+	int32_t mode = 0777;
+	int flags = O_RDWR | O_CREAT;
+
+	attrs.mq_msgsize = MESSAGE_SIZE;
+	attrs.mq_maxmsg = MESG_COUNT_PERMQ;
+
+	not.sigev_notify = SIGEV_THREAD;
+	not.sigev_value.sival_int = (int)pthread_self();
+	not.sigev_notify_function = notify_function_thread;
+	not.sigev_notify_attributes = NULL;
```
Similar - again, not blocking, so if this does need a respin for any reason, it might be worth it.
```suggestion
	mqd_t mqd;
	struct mq_attr attrs = {
		.mq_msgsize = MESSAGE_SIZE,
		.mq_maxmsg = MESG_COUNT_PERMQ,
	};
	struct sigevent not = {
		.sigev_notify = SIGEV_THREAD,
		.sigev_value.sival_int = (int)pthread_self(),
		.sigev_notify_function = notify_function_thread,
	};
	int32_t mode = 0777;
	int flags = O_RDWR | O_CREAT;
```

cfriedt commented on tests/posix/common/src/mqueue.c:
```diff
@@ -97,4 +96,205 @@ ZTEST(mqueue, test_mqueue)
+
+ZTEST(mqueue, test_mqueue_notify_basic)
+{
+	mqd_t mqd;
+	struct mq_attr attrs;
+	struct sigevent not;
+	int32_t mode = 0777;
+	int flags = O_RDWR | O_CREAT;
+
+	attrs.mq_msgsize = MESSAGE_SIZE;
+	attrs.mq_maxmsg = MESG_COUNT_PERMQ;
+
+	not.sigev_notify = SIGEV_NONE;
+	not.sigev_value.sival_ptr = (void *)&notification_executed;
+	not.sigev_notify_function = notify_function_basic;
+	not.sigev_notify_attributes = NULL;
```
Same - just a suggestion, not blocking, so you can ignore if there is no respin required.
```suggestion
	mqd_t mqd;
	struct mq_attr attrs = {
		.mq_msgsize = MESSAGE_SIZE,
		.mq_maxmsg = MESG_COUNT_PERMQ,
	};
	struct sigevent not = {
		.sigev_notify = SIGEV_NONE,
		.sigev_value.sival_ptr = (void *)&notification_executed,
		.sigev_notify_function = notify_function_basic,
	};
	int32_t mode = 0777;
	int flags = O_RDWR | O_CREAT;
```

cfriedt commented on tests/posix/common/src/mqueue.c:
```diff
@@ -97,4 +96,205 @@ ZTEST(mqueue, test_mqueue)
 	zassert_false(mq_unlink(queue), "Not able to unlink Queue");
 }
 
+bool notification_executed;
```
Not blocking either, so only if a respin is needed.
```suggestion
static bool notification_executed;
```

cfriedt commented on lib/posix/mqueue.c:
```diff
@@ -341,6 +347,60 @@ int mq_setattr(mqd_t mqdes, const struct mq_attr *mqstat,
+	mqueue_object *mqueue = (mqueue_object *)arg;
+	struct sigevent *sevp = &mqueue->not;
+
+	pthread_detach(pthread_self());
```
Aside from calling `pthread_detach()`, I think it might be necessary to set the cancel behaviour to asynchronous. Otherwise, calling `pthread_cancel()` might have no (immediate) effect.

https://github.com/zephyrproject-rtos/zephyr/pull/67871 for reference, since there was a similar recent bug in `timer.c`

cfriedt commented on tests/posix/common/src/mqueue.c:
```diff
@@ -97,4 +96,205 @@ ZTEST(mqueue, test_mqueue)
+
+ZTEST(mqueue, test_mqueue_notify_non_empty_queue)
+{
+	mqd_t mqd;
+	struct mq_attr attrs;
+	struct sigevent not;
+	int32_t mode = 0777;
+	int flags = O_RDWR | O_CREAT;
+
+	attrs.mq_msgsize = MESSAGE_SIZE;
+	attrs.mq_maxmsg = MESG_COUNT_PERMQ;
+
+	not.sigev_notify = SIGEV_NONE;
+	not.sigev_value.sival_ptr = (void *)&notification_executed;
+	not.sigev_notify_function = notify_function_basic;
+	not.sigev_notify_attributes = NULL;
```
Similar - would probably be a good idea to initialize at declaration time (that way unspecified members are zero-initialized).

cfriedt commented on lib/posix/mqueue.c:
```diff
@@ -341,6 +347,82 @@ int mq_setattr(mqd_t mqdes, const struct mq_attr *mqstat,
+		return -1;
+	}
+	if (notification->sigev_notify_attributes != NULL) {
+		ret = pthread_attr_getdetachstate(notification->sigev_notify_attributes,
+						  &detachstate);
+		if (ret != 0) {
+			errno = ret;
+			return -1;
+		}
+		if (detachstate != PTHREAD_CREATE_DETACHED) {
```
No need to check what the attr detachstate is before setting it
```suggestion
		ret = pthread_attr_getdetachstate(notification->sigev_notify_attributes,
						  &detachstate);
		if (ret != 0) {
			errno = ret;
			return -1;
		}
		if (detachstate != PTHREAD_CREATE_DETACHED) {
```

awojasinski replied:
That's the way `posix/timer.c` do it so maybe it should be changed there as well?

cfriedt commented on lib/posix/mqueue.c:
```diff
@@ -341,6 +347,82 @@ int mq_setattr(mqd_t mqdes, const struct mq_attr *mqstat,
+				errno = ret;
+				return -1;
+			}
+		}
```
```suggestion
```

cfriedt reviewed (commented):
No need to check the attr detachstate before setting it - otherwise basically done 👍

ycsin commented on tests/posix/common/src/mqueue.c:
```diff
@@ -97,4 +96,201 @@ ZTEST(mqueue, test_mqueue)
+	mqd = mq_open(queue, flags, mode, &attrs);
+
+	zassert_false(mq_send(mqd, send_data, MESSAGE_SIZE, 0),
+		      "Not able to send message");
```
This and others, to be consistent with other lines that used "Unable"
```suggestion
		      "Unable to send message");
```

cfriedt reviewed (dismissed):
LGTM - thanks @awojasinski 👍

ycsin commented on tests/posix/common/src/mqueue.c:
```diff
@@ -97,4 +96,185 @@ ZTEST(mqueue, test_mqueue)
+
+	usleep(USEC_PER_MSEC * 10U);
+
+	zassert_true(notification_executed,
+		     "Notification not triggered.");
```
nit: this line is different from the other 2
```suggestion
	zassert_true(notification_executed, "Notification not triggered.");
```

awojasinski replied:
fixed

ycsin reviewed (approved):
![](https://i.kym-cdn.com/entries/icons/original/000/021/267/swedish_chef.jpg)

cfriedt commented:
@awojasinski - if you're feeling generous, we could also use [mq_timedreceive()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/mq_timedreceive.html) and  [mq_timedsend()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/mq_timedsend.html) for the [_POSIX_TIMEOUTS](https://docs.zephyrproject.org/latest/services/portability/posix/option_groups/index.html#posix-timeouts) subprofiling option 😃

awojasinski commented:
> @awojasinski - if you're feeling generous, we could also use [mq_timedreceive()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/mq_timedreceive.html) and [mq_timedsend()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/mq_timedsend.html) for the [_POSIX_TIMEOUTS](https://docs.zephyrproject.org/latest/services/portability/posix/option_groups/index.html#posix-timeouts) subprofiling option 😃

@cfriedt sure I'll take a look in my spare time
